File: Data Structures & Algorithms/count-connected-components/submission-6.cs
public class Solution {
    public int CountComponents(int n, int[][] edges) {
        var adjList = new List<int>[n];
        for (int i = 0; i < n; i++) adjList[i] = new List<int>();
        foreach (int[] edge in edges) {
            int n1 = edge[0], n2 = edge[1];
            adjList[n1].Add(n2);
            adjList[n2].Add(n1);
        }

        var visited = new HashSet<int>();
        void Dfs(int vert) {
            if (visited.Contains(vert)) return;

            visited.Add(vert);
            foreach (int connectedVert in adjList[vert]) Dfs(connectedVert);
        }

        int count = 0;
        for (int vert = 0; vert < n; vert++) {
            if (!visited.Contains(vert)) {
                count++;
                Dfs(vert);
            }
        }
        return count;
    }
}

File: Data Structures & Algorithms/valid-tree/submission-6.cpp
class Solution {
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        if (ssize(edges) != n - 1) return false;

        vector<vector<int>> adj_list(n);
        for (const vector<int>& edge : edges) {
            adj_list[edge[0]].push_back(edge[1]);
            adj_list[edge[1]].push_back(edge[0]);
        }

        unordered_set<int> visited;
        function<bool(int node, int parent)> dfs = [&](int node, int parent) -> bool {
            if (visited.contains(node)) return false;

            visited.insert(node);
            for (int neighbor : adj_list[node]) {
                if (parent == neighbor) continue;
                if (!dfs(neighbor, node)) return false;
            }

            return true;
        };

        return dfs(0, -1) && ssize(visited) == n;
    }
};
